---
title: |
  ![](UF.png){width=10%}

  Genomic whole association study (GWAS
  )
author: 
  - Marcio Resende, professor^[University of Florida, mresende@ufl.edu]
  
  - Marco Antonio Peixoto, Post doc^[University of Florida, deamorimpeixotom@ufl.edu]
  
output: 
  html_document:
    toc: true
    toc_float: FALSE
    theme: flatly
    highlight: haddock
bibliography: references.bib
---

```{=html}
<style type="text/css">
  body{
  font-size: 14pt;
}
</style>
```

## Introduction

This is an script developed using a population simulated in AlphaSimR package [@gaynor2021alphasimr]. The GWAS models implemented came from four different models: 

i. Unadjusted markers
ii. Adjusted by Q
iii. Adjusted by K
iv. Adjusted by K+Q

## Loading the packages for the GWAS

```{r eval=TRUE}
#Example GWAS analysis for Q+K models
library(AlphaSimR) 
library(rrBLUP) #For K and Q+K GWAS
library(qqman) #For Manhattan and Q-Q plots
library(ggplot2)
library('ggfortify')

```

## Creating the population for the GWAS


### 1. Creating the base genome

```{r eval=TRUE}
# Global parameters
nQtlPerChr = 10 #Per pair of chromosome
nSnpPerChr = 500 # must be > nQtlPerChr

# Simulate mapping population
set.seed(18556255)
FOUNDERPOP = runMacs(nInd=1000,
                     nChr=10,
                     segSites=nSnpPerChr,
                     split=20)

# Simulate the trait of interest
SP = SimParam$
  new(FOUNDERPOP)$
  restrSegSites(minSnpPerChr=nSnpPerChr,
                minQtlPerChr=nSnpPerChr,
                overlap=TRUE)$
  addTraitA(nQtlPerChr,gamma=TRUE)$
  addSnpChip(nSnpPerChr)$
  setVarE(H2=0.8) # Trait heritability

pop = newPop(FOUNDERPOP)

```

### 2. Creating a population structure

In a way to create a population structure, we implemented two steps. The first was to split the base genome into two. We used above the argument `split=20`, which represents that the base genome was splited in two 20 generations ago. The second, is going to be the cycles of selections below. as we have two groups, we will split the base population (pop) into two (one from 1:500, and the second from 501:1000). With the first population (PopA) going to 4 cycles of selections and the second population (PopB) facing 4 cycles of selection. As they differ, it is likely that the allele frequency of the two populations would be different themselves.

```{r eval=TRUE}

#Create 2 sub populations
popA = pop[1:500]
popB = pop[501:1000]

#Cycles of selection
for(i in 1:6){
  popA = selectCross(popA,
                     nInd=10,
                     nCrosses=50,
                     nProgeny=10)
  if(i<4){
    popB = selectCross(popB,
                       nInd=10,
                       nCrosses=50,
                       nProgeny=10)
  }
}

#Combining both pops
pop = c(popA,popB)

```



## PCA and population genotypes


In this step we will extract the SNPs from the core population and we will create a PCA to explore the structure in the data.

```{r eval=TRUE}
#Format genotypes for rrBLUP
rawGeno = pullSnpGeno(pop)
freq = colMeans(rawGeno)/2
MAF = apply(cbind(freq,1-freq),1,min)

#Ploting PCA
pca_res <- prcomp(rawGeno)

autoplot(pca_res)



```

## Organizing the data for the GWAS 

For the GWAS we need to get the data. We will need dosage matrix, phenotypes values for the trait of interest, and the QTL positions.

```{r eval=TRUE}

# Genotypes/SNP matrix/ dosage matrix
rawGeno = rawGeno-1
geno = as.data.frame(t(rawGeno))
geno = data.frame(snp=row.names(geno),
                  chr=rep(1:10,each=nSnpPerChr),
                  pos=rep(1:nSnpPerChr,10),
                  geno)
geno[1:10, 1:10]

#Create "pheno" data.frames for rrBLUP
pheno = data.frame(gid=names(geno)[-(1:3)],
                   trait1=pop@pheno[,1])

phenoQ = data.frame(gid=pheno$gid,
                    subPop=factor(rep(c("a","b"),each=500)),
                    trait1=pheno$trait1)

#Find QTL locations within SNP chip
qtl = paste0("SNP_",
             SP$traits[[1]]@lociLoc+
               rep(0:9*nSnpPerChr,each=nQtlPerChr))

```

## Model 1 - GWAS with no adjustament

```{r eval=TRUE}

##GWAS
#No adjustments----
model0 = data.frame(snp=geno$snp,
                    chr=geno$chr,
                    pos=geno$pos,
                    trait1=rep(NA_real_,10*nSnpPerChr))

for(i in 1:(10*nSnpPerChr)){
  if(MAF[i]>=0.05){
    #Fit linear model with aov and extract p-value
    tmp = summary(aov(pheno$trait1~rawGeno[,i]))[[1]][1,5]
    #Check for bad p-values
    #Check for markers confounded with structure
    if(is.na(tmp)){ 
      model0$trait1[i] = 1
    }else{
      model0$trait1[i] = tmp
    }
  }else{
    model0$trait1[i] = 1
  }
}
#Account for p-value=0
model0$trait1[model0$trait1==0] = 1e-300

```

## Model 2 - GWAS correcting with Q

```{r eval=TRUE}
#Adjust for population structure (Q)----
modelQ = model0
for(i in 1:(10*nSnpPerChr)){
  if(MAF[i]>=0.05){
    #Fit linear model with aov and extract p-value
    tmp = summary(aov(pheno$trait1~phenoQ$subPop+rawGeno[,i]))
    tmp = tmp[[1]][2,5]
    #Check for markers confounded with structure
    if(is.na(tmp)){ 
      modelQ$trait1[i] = 1
    }else{
      modelQ$trait1[i] = tmp
    }
  }else{
    modelQ$trait1[i] = 1
  }
}
#Account for p-value=0
modelQ$trait1[modelQ$trait1==0] = 1e-300

```

## Model 3 - GWAS correcting with Kinship matrix (K)

```{r eval=TRUE}
#Adjust for kinship (K)----
modelK = GWAS(pheno,geno,plot=FALSE)
modelK$trait1 = 10^(-modelK$trait1) #Revert to p-value

```

## Model 4 - GWAS correcting with both, Q and K

```{r eval=TRUE}
#Adjust for structure and kinship (Q+K)----
modelQK = GWAS(phenoQ,geno,fixed="subPop",plot=FALSE)
modelQK$trait1 = 10^(-modelQK$trait1) #Revert to p-value

```

## Manhattan plot
```{r eval=TRUE}
#Manhattan plots
op = par(mfrow=c(2,2),mai=c(.9,.9,0.9,0.9),
         mar=c(2.5,2.5,1,1)+0.1,
         mgp = c(1.5,0.5,0))
manhattan(model0,chr="chr",bp="pos",p="trait1",snp="snp",
          highlight=qtl,main="Unadjusted")
manhattan(modelQ,chr="chr",bp="pos",p="trait1",snp="snp",
          highlight=qtl,main="Adjusted for Q")
manhattan(modelK,chr="chr",bp="pos",p="trait1",snp="snp",
          highlight=qtl,main="Adjusted for K")
manhattan(modelQK,chr="chr",bp="pos",p="trait1",snp="snp",
          highlight=qtl,main="Adjusted for Q+K")
par(op)


```
## QQ-plot
```{r eval=TRUE}
#Q-Q plots
op = par(mfrow=c(2,2),mai=c(.9,.9,0.9,0.9),
         mar=c(2.5,2.5,1,1)+0.1,
         mgp = c(1.5,0.5,0))
qq(model0$trait1,main="Unadjusted")
qq(modelQ$trait1,main="Adjusted for Q")
qq(modelK$trait1,main="Adjusted for K")
qq(modelQK$trait1,main="Adjusted for Q+K")
par(op)

```

# 6. References

::: {#refs}
:::
