---
title: |
  ![](UF.jpg){width=10%}

  Cross validation scheme 2 (CV2) in BGLR package
author: 
 
  - Marco Antonio Peixoto, Post doc^[University of Florida, deamorimpeixotom@ufl.edu]
  
output: 
  html_document:
    toc: false
    toc_float: false
    theme: flatly
    highlight: haddock
bibliography: references.bib
csl: apa.csl   
---

```{=html}
<style type="text/css">
  body{
  font-size: 12pt;
}
</style>
```


<br>


## 1. Genomic selection in a CV2 scheme

In this file I show how to implement a **GBLUP** model in a CV2 scheme using BGLR [@perez2022multitrait] in a Single- and Multi-trait framework.

To perform the analyses, we will need the following packages:

```{r echo=TRUE, warning=FALSE,eval=FALSE}
#--- 1. Cleaning
rm(list=ls())

#--- 2. Packages
require(AGHmatrix)
require(BGLR)
require(tidyverse)
require(dplyr)
require(ggplot2)
```

<br>

## 2. Datasets 

This dataset was simulated using the coalescent theory implemented in $AlphasimR$ package [@gaynor2021alphasimr]. IT mimics an evaluation of 500 maize genotypes, in four locations, and four traits were measured. Also, a set of 3000 single nucleotide polymorphism (SNPs) were randomly sampled through the 10 pair of chromosomes.

**Phenotypic data**  
We generated BLUEs (best linear unbiased estimation) for each trait, and it can be loaded from *Phenotypes.csv* file.

```{r echo=TRUE, eval=TRUE}
#--- 1. Loading the data
Phenotypes = read.csv("Phenotypes.csv")

#--- 2. As factor
Phenotypes$Genotype = Phenotypes$Genotype %>% as.factor
Phenotypes$Env = Phenotypes$Env %>% as.factor

```


<br>

**Genomic data**  
The genomic data (3000 SNP markers) was coded as 0,1,2 for aa,Aa,AA, respectively. The dataset for all 500 genotypes can be loaded from **Genotypes.csv**.

```{r echo=FALSE, eval=TRUE}
#--- 1. SNPs
Genotypes = as.matrix(read.csv("Genotypes.csv"))

#--- 2. Names
rownames(Genotypes)= unique(Phenotypes$Genotype)

```

<br>

**Creating G Matrix**

Here, we will compose the additive relationship matrix (*G* matrix) following the propositions made by @vanraden2008efficient.

```{r echo=FALSE, eval=TRUE}
#--- 1. Additive matrix
G.Mat = AGHmatrix::Gmatrix(Genotypes) 

```

<br>

## 3. Statistical model


**Single-trait model**

The **GBLUP** model in a singletrait framework is given as follows:

$$
y = 1u + Za + e
$$

where $y$ is the vector of BLUEs data for the target trait; $u$ represents the mean (fixed); $a$ is the vector of genotype effects (assumed to be random), where $a \sim N(0,G\sigma^2_a)$, $G$ is a matrix that describes the genomic similarities between pair of genotypes and $\sigma^2_a$ represents the genetic variance; and $e$ is the vector of residuals (random), where $e\sim N(0,I\sigma^2_{res})$, $\sigma^2_{res}$ is the residual variance. The letter $Z$ refers to the incidence matrix for $a$.

<br>


**Multi-trait model**

The **GBLUP** model in a multitrait framework is given as follows:

$$
y = 1u + Za + e
$$

where $y$ is the matrix of BLUEs data for the target traits; $u$ represents the mean (fixed); $a$ is the matrix of genotype effects (assumed to be random), where $a \sim N(0,\sum_a\otimes G)$, $G$ is a matrix that describes the genomic similarities between pair of genotypes and $\sum_a$ represents the genetic covariance matrix; and $e$ is the  residuals (random), where $e\sim N(0,\sum_{res}\otimes I)$, $\sum_{res}$ is the residual covariance matrix. The letter $Z$ refers to the incidence matrix for $a$. 

<br>

## 4. Single-trait GBLUP using BGLR() function

For the model implementation we use three different strategies. The first strategy is the prediction a single trait model where I used the information from the trait to predict individuals in the same environment. 

So, we will use all trait 1 over four environments.

```{r echo=TRUE, eval=TRUE}
#--- 1. Organizing the phenotypic data

Y<- Phenotypes[,c(1,2,3)] %>%  
  pivot_wider(., names_from = 'Env', values_from = 'Trait1') 

#--- 2. Scale the data
y = scale(Y[,-1])


```


#### **Single-trait model**

Implementing the model using the function BGLR()


```{r echo=TRUE, eval=TRUE}

#--- 1. Setting NAs for the individuals
set.seed(524521)
yNA <-y
yNA[sample(1:dim(Y)[1],100),1] <- NA
yNA[sample(1:dim(Y)[1],100),2] <- NA
yNA[sample(1:dim(Y)[1],100),3] <- NA
yNA[sample(1:dim(Y)[1],100),4] <- NA



#--- 2. Preparing the single trait model
  y1NA <- yNA[,1] #Trait 1
  y2NA <- yNA[,2] #Trait 2
  y3NA <- yNA[,3] #Trait 3
  y4NA <- yNA[,4] #Trait 4

#--- 3. ETA and model for each trait
  ETA <- list(A=list(K=G.Mat, model='RKHS'))
  fm1 <- BGLR(y1NA,ETA=ETA,nIter=1200,burnIn=120,verbose=FALSE)
  fm2 <- BGLR(y2NA,ETA=ETA,nIter=1200,burnIn=120,verbose=FALSE)
  fm3 <- BGLR(y3NA,ETA=ETA,nIter=1200,burnIn=120,verbose=FALSE)
  fm4 <- BGLR(y4NA,ETA=ETA,nIter=1200,burnIn=120,verbose=FALSE)

#--- 4. Correlation for single trait
  COR.ST1 <- cor(fm1$yHat[tst1],y[tst1,1])
  COR.ST2 <- cor(fm2$yHat[tst2],y[tst2,2])
  COR.ST3 <- cor(fm3$yHat[tst3],y[tst3,3])
  COR.ST4 <- cor(fm4$yHat[tst4],y[tst4,4])
  

```


#### **Multi trait model**

The same information is used with several traits at once. In this second scenario, we used the function Multitrait() from BGLR. The prediction in this case is for the partial information via associated traits.



```{r echo=TRUE, eval=TRUE}

#--- 1. Set to NA the
set.seed(524521)
yNA <-y
yNA[sample(1:dim(Y)[1],100),1] <- NA
yNA[sample(1:dim(Y)[1],100),2] <- NA
yNA[sample(1:dim(Y)[1],100),3] <- NA
yNA[sample(1:dim(Y)[1],100),4] <- NA

#--- 2. Model
fm <- Multitrait(y = yNA,
                 ETA = list(A=list(K = G.Mat, model='RKHS', Cov = list(type = 'UN'))), 
                 nIter = 1200, 
                 burnIn = 120, 
                 thin = 5, 
                 verbose = FALSE)

#--- 3. Extracting the estimated parameters
YHatInt <- fm$ETAHat

#--- 4. Predictive correlation
test<-is.na(yNA)
tst1<-test[,1];tst2<-test[,2]<-tst3<-test[,3];tst4<-test[,4]
COR.T1<-cor(YHatInt[tst1,1],y[tst1,1])
COR.T2<-cor(YHatInt[tst2,2],y[tst2,2])
COR.T3<-cor(YHatInt[tst3,3],y[tst3,3])
COR.T4<-cor(YHatInt[tst4,4],y[tst4,4])

#--- 5. Extracting estimates of variance parameters
(COR.RES<-fm$resCov$R) # residual covariance matrix
(COR.u<-fm$ETA[[1]]$Cov$Omega) # genetic covariance matrix UE

#--- 6. Trait correlation  
cov2cor(CORB.u)
```

<br>

#### **Correlation for the models (MT and ST)**  

```{r, echo=TRUE, eval=TRUE}  

#--- 1. Correlation output from both models
CorCV2 <- matrix(NA,nrow=4,ncol=2)
colnames(CorCV2) <- c('MT','ST')

  # MTM
  CorCV2[1,1] <- COR.T1
  CorCV2[2,1] <- COR.T2
  CorCV2[3,1] <- COR.T3
  CorCV2[4,1] <- COR.T4
 
  # STM
  CorCV2[1,2] <- COR.ST1
  CorCV2[2,2] <- COR.ST2
  CorCV2[3,2] <- COR.ST3
  CorCV2[4,2] <- COR.ST4

#--- 2. Comparison table

CorCV2

```







## References
